field/composite.go

```diff
@@ -117,55 +119,75 @@ func (f *Composite) SetData(data interface{}) error {
 
 // Pack deserialises data held by the receiver (via SetData)
 // into bytes and returns an error on failure.
-func (f *Composite) Pack() ([]byte, error) {
-	packed, err := f.pack()
+func (f *Composite) WriteTo(w io.Writer) (n int, err error) {
+	packed := bytes.NewBuffer([]byte{})
+	err = f.pack(packed)
 	if err != nil {
-		return nil, err
+		return 0, err
+	}
+
+	packedLength, err := f.spec.Pref.EncodeLength(f.spec.Length, packed.Len())
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode length: %v", err)
 	}
 
-	packedLength, err := f.spec.Pref.EncodeLength(f.spec.Length, len(packed))
+	m, err := w.Write(packedLength)
+	if err != nil {
+		return m, fmt.Errorf("writing packed length: %v", err)
+	}
+
+	n += m
+
+	m, err = w.Write(packed.Bytes())
 	if err != nil {
-		return nil, fmt.Errorf("failed to encode length: %v", err)
+		return m, fmt.Errorf("writing packed field: %v", err)
 	}
 
-	return append(packedLength, packed...), nil
+	n += m
+
+	return n, nil
 }
 
-// Unpack takes in a byte array and serializes them into the receiver's
-// subfields. An offset (unit depends on encoding and prefix values) is
-// returned on success. A non-nil error is returned on failure.
-func (f *Composite) Unpack(data []byte) (int, error) {
-	dataLen, err := f.spec.Pref.DecodeLength(f.spec.Length, data)
+// ReadFrom takes in an io.Reader and reads data from it and then serializes
+// them into the receiver's subfields. An offset (unit depends on encoding and
+// prefix values) is returned on success. A non-nil error is returned on
+// failure.
+func (f *Composite) ReadFrom(r io.Reader) (int, error) {
+	dataLen, err := f.spec.Pref.ReadLength(f.spec.Length, r)
 	if err != nil {
 		return 0, fmt.Errorf("failed to decode length: %v", err)
 	}
-	offset := f.spec.Pref.Length()
-
-	read, err := f.unpack(data[offset:])
+	read, err := f.unpack(r, dataLen)
 	if err != nil {
 		return 0, err
 	}
 	if dataLen != read {
 		return 0, fmt.Errorf("data length: %v does not match aggregate data read from decoded subfields: %v", dataLen, read)
 	}
 
-	return offset + read, nil
+	return f.spec.Pref.Length() + read, nil
 }
 
 // SetBytes iterates over the receiver's subfields and unpacks them.
 // Data passed into this method must consist of the necessary information to
-// pack all subfields in full. However, unlike Unpack(), it requires the
+// pack all subfields in full. However, unlike ReadFrom(), it requires the
 // aggregate length of the subfields not to be encoded in the prefix.
 func (f *Composite) SetBytes(data []byte) error {
-	_, err := f.unpack(data)
+	_, err := f.unpack(bytes.NewReader(data), len(data))
 	return err
 }
 
 // Bytes iterates over the receiver's subfields and packs them. The result
 // does not incorporate the encoded aggregate length of the subfields in the
 // prefix.
 func (f *Composite) Bytes() ([]byte, error) {
-	return f.pack()
+	buf := bytes.NewBuffer([]byte{})
```

Comment:
```suggestion
	var buf bytes.Buffer
```

Reply:
If you do need to allocate memory ahead of time there's `Buffer.Grow(int)` as well. 

https://golang.org/pkg/bytes/#Buffer.Grow